src/Infrastructure/Process/FileCopier.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace PhpTuf\ComposerStager\Infrastructure\Process;
+
+use PhpTuf\ComposerStager\Exception\LogicException;
+use PhpTuf\ComposerStager\Exception\ProcessFailedException;
+use PhpTuf\ComposerStager\Infrastructure\Process\Runner\RsyncRunner;
+
+/**
+ * @internal
+ */
+class FileCopier
+{
+    /**
+     * @var \PhpTuf\ComposerStager\Infrastructure\Process\Runner\RsyncRunner
+     */
+    private $rsync;
+
+    public function __construct(RsyncRunner $rsync)
+    {
+        $this->rsync = $rsync;
+    }
+
+    /**
+     * @param string $from
+     * @param string $to
+     * @param string[] $exclusions Paths to exclude, relative to the "from" path.
+     * @param callable|null $callback An optional PHP callback to run whenever
+     *   there is some output available on STDOUT or STDERR.
+     *
+     * @see https://symfony.com/doc/current/components/process.html#running-processes-asynchronously
+     *
+     * @throws \PhpTuf\ComposerStager\Exception\ProcessFailedException
+     */
+    public function copy(string $from, string $to, array $exclusions = [], ?callable $callback = null): void
+    {
+        $command = [
+            '--recursive',
+            // The "--links" option is added to "copy symlinks as symlinks",
+            // which is important particularly for files in vendor/bin.
+            '--links',
+            '--verbose',
+        ];
+
+        foreach ($exclusions as $exclusion) {
+            $command[] = '--exclude=' . $exclusion;
+        }
+
+        // A trailing slash is added to the "from" directory so the CONTENTS of
```

Comment:
Thanks for this comment; it makes it clear why the extra slash is not arbitrary. Otherwise it would be too easy to mistake this as a stylistic thing. I wonder: do we (or will we) have functional test coverage to prove that the contents of the directory are copied properly?

Reply:
We don't have functional tests _yet_. They wouldn't have been feasible without the `Beginner`. But it's my next priority to lay a foundation for functional testing and start writing tests.